### PR TITLE
ignore im2col if not necessary in conv 1 * 1

### PR DESCRIPTION
**Pull request by NHZlX**
ignore im2col when this happens in conv:
- filteHeight ==  1 && filterWidth == 1
- paddingH == 0 && paddingW == 0
- strideH == 1 && strideW == 1

I did a test of mobilenet forward process on Developing machine(Intel(R) Xeon(R) CPU E5-2620 v2 @ 2.10GHz), the average cost time on 100 times forward pass of one batch is about 15 ms faster than the normal one.

|withIm2col|average time|
|:--:|:--:|
|True|0.2359s|
|Fase|0.22018s|

### CODE REVIEW

File: paddle/function/GemmConvOp.cpp [hunk 1]
```diff
@@ -66,16 +66,23 @@ class GemmConvFunction : public ConvFunctionBase {
     real* inputData = inputs[0].data<real>();
     real* filterData = inputs[1].data<real>();
     real* outputData = outputs[0].data<real>();
+    bool skipIm2col = isSkipIm2col(filter);
```
Comment (hedaoyuan) [inline paddle/function/GemmConvOp.cpp]: `skipIm2col`换成`needIm2col`是否更好理解?

File: paddle/function/GemmConvOp.cpp [hunk 2]
```diff
@@ -66,16 +66,23 @@ class GemmConvFunction : public ConvFunctionBase {
...
-    TensorShape colShape = TensorShape({inputChannels / groups_,
-                                        filterHeight,
-                                        filterWidth,
-                                        outputHeight,
-                                        outputWidth});
 
-    resizeBuffer<Device>(colShape.getElements());
-    real* colData = reinterpret_cast<real*>(memory_->getBuf());
+    TensorShape colShape;
+    real *colBuffer, *colData = NULL;
```
Comment (hedaoyuan) [inline paddle/function/GemmConvOp.cpp]: 一行定义一个变量。实际上只需要一个colData变量就够了。
97行写成
```
if (needIm2col) {
  im2col(colData);
} else {
  colData = inputData + g * inputOffset;
}
```

File: paddle/function/GemmConvOp.cpp [hunk 3]
```diff
@@ -182,6 +200,12 @@ class GemmConvGradInputFunction : public ConvFunctionBase {
         int K = outputChannels / groups_;
         int N = outputHeight * outputWidth;
         int M = inputChannels / groups_ * filterHeight * filterWidth;
+        colBuffer = colData;
+        real scale = 0.0f;
```
Comment (hedaoyuan) [inline paddle/function/GemmConvOp.cpp]: 203-204写到if(skipIm2col)...else里面吧

File: paddle/function/GemmConvOp.cpp (hunk 2 above)
Comment (NHZlX) [inline paddle/function/GemmConvOp.cpp]: Done

File: paddle/function/GemmConvOp.cpp (hunk 1 above)
Comment (NHZlX) [inline paddle/function/GemmConvOp.cpp]: ok, sounds great.

File: paddle/function/GemmConvOp.cpp (hunk 3 above)
Comment (NHZlX) [inline paddle/function/GemmConvOp.cpp]: read scale = 0.0  没必要放在if else 中吧

Comment (hedaoyuan) [inline paddle/function/GemmConvOp.cpp]: 嗯，把colBuffer去掉就行了。

**Pull request merged**